lib/vrl/stdlib/src/parse_aws_cloudtrail_logs.rs

```diff
@@ -0,0 +1,1063 @@
+        let mut value = BTreeMap::<String, Value>::new();
+
+        value.insert("average".to_owned(), statistics.average.into());
+
+        value.into()
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all(deserialize = "camelCase"))]
+struct AwsCloudTrailLogsRecordInsightDetailsInsightContextAttributions {
+    attribute: AwsCloudTrailLogsRecordInsightDetailsInsightContextAttributionsAttribute,
+    insight: Vec<AwsCloudTrailLogsRecordInsightDetailsInsightContextAttributionsInsight>,
+    baseline: Vec<AwsCloudTrailLogsRecordInsightDetailsInsightContextAttributionsBaseline>,
+}
+
+impl From<AwsCloudTrailLogsRecordInsightDetailsInsightContextAttributions> for Value {
+    fn from(attributions: AwsCloudTrailLogsRecordInsightDetailsInsightContextAttributions) -> Self {
+        let mut value = BTreeMap::<String, Value>::new();
+
+        value.insert("attribute".to_owned(), attributions.attribute.into());
+        value.insert("insight".to_owned(), attributions.insight.into());
+        value.insert("baseline".to_owned(), attributions.baseline.into());
+
+        value.into()
+    }
+}
+
+#[derive(Clone, Copy, Debug, Serialize, Deserialize)]
+#[serde(rename_all(deserialize = "camelCase"))]
+enum AwsCloudTrailLogsRecordInsightDetailsInsightContextAttributionsAttribute {
+    UserIdentityArn,
+    UserAgent,
+    ErrorCode,
+}
+
+impl From<AwsCloudTrailLogsRecordInsightDetailsInsightContextAttributionsAttribute> for Value {
+    fn from(
+        attribute: AwsCloudTrailLogsRecordInsightDetailsInsightContextAttributionsAttribute,
+    ) -> Self {
+        format!("{:?}", attribute).into()
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all(deserialize = "camelCase"))]
+struct AwsCloudTrailLogsRecordInsightDetailsInsightContextAttributionsInsight {
+    value: String,
+    average: f64,
+}
+
+impl From<AwsCloudTrailLogsRecordInsightDetailsInsightContextAttributionsInsight> for Value {
+    fn from(
+        insight: AwsCloudTrailLogsRecordInsightDetailsInsightContextAttributionsInsight,
+    ) -> Self {
+        let mut value = BTreeMap::<String, Value>::new();
+
+        value.insert("value".to_owned(), insight.value.into());
+        value.insert("average".to_owned(), insight.average.into());
+
+        value.into()
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all(deserialize = "camelCase"))]
+struct AwsCloudTrailLogsRecordInsightDetailsInsightContextAttributionsBaseline {
+    value: String,
+    average: f64,
+}
+
+impl From<AwsCloudTrailLogsRecordInsightDetailsInsightContextAttributionsBaseline> for Value {
+    fn from(
+        baseline: AwsCloudTrailLogsRecordInsightDetailsInsightContextAttributionsBaseline,
+    ) -> Self {
+        let mut value = BTreeMap::<String, Value>::new();
+
+        value.insert("value".to_owned(), baseline.value.into());
+        value.insert("average".to_owned(), baseline.average.into());
+
+        value.into()
+    }
+}
```

Comment:
There's a lot of "dumb" conversion code here to produce a VRL `Value`.

The alternative would be to
1. deserialize to `AwsCloudTrailLogs`
2. serialize to JSON
3. deserialze VRL `Value`

But those string operations are certainly more costly than the manual conversion code.

Reply:
...continuing https://github.com/timberio/vector/pull/7488#issuecomment-844370504:

The boilerplate isn't there to deserialize, but to produce a vrl `Value`, since (to my knowledge) there exist is no direct conversion between the deserialized `AwsCloudTrailLogs` to `Value` without going through JSON serialization.

Alternatively, I could first deserialize to `serde_json::Value` and then "hand-pick" and rename the desired fields when moving to vrl `Value`. However, having those type-safe structs seemed to be the better alternative to me.

If you have another approach that I can't see right now, I'd be happy to omit the manual code 😄

Reply:
> Alternatively, I could first deserialize to `serde_json::Value` and then "hand-pick" and rename the desired fields when moving to vrl `Value`. However, having those type-safe structs seemed to be the better alternative to me.

Yeah, this is what I was thinking, that you could just deserialize to `serde_json::Value` and then convert that directly to a `vrl::Value` via the `From` implementation you added here. I think the type-safe deserializing is useful if we were going to be modifying the data before converting it to a `vrl::Value`, but I'm not seeing us do that anywhere so I'm not sure what it is buying us. I could definitely be missing something though.

Reply:
For what it's worth, we do the same in `parse_aws_cloudwatch_log_subscription_message`:

https://github.com/timberio/vector/blob/1e54aca6970a3f141749bd56d8cd8c98ba6e3fab/lib/vrl/stdlib/src/parse_aws_cloudwatch_log_subscription_message.rs#L75-L89

In that case the destructuring/mapping to `Value` is more trivial (no optional values, only one nested struct), so we don't need a `From` implementation there.

What I like about the structs though is that you have a declarative description of what the format is, instead of needing to read the transformation code to understand what it ends up like.

Reply:
🤔 true. I'm ok if you want to leave it; it just seemed like it wasn't strictly needed. We may want to do additional transformations in the future anyway.